app/models/event_log.rb

```diff
@@ -0,0 +1,5 @@
+# frozen_string_literal: true
+
+class EventLog < ApplicationRecord
+  belongs_to :owner, polymorphic: true
+end
```

Comment:
Do we expect to show the events to anyone, like in a timeline or something? At the moment I don't think we are doing anything with them, other than storing them, and for storing object versions over time papertrail is likely to do a much better job than we are.

I guess that is really the question: are we sure we care about the history of partnership statuses enough to create our own events for them?

Reply:
I agree that papertrail is great in what it's doing - it maintains a list of versions of given database table. Event log on the other hand does not track the state of the records and only records explicit domain events forming a subset of transaction log. The advantage is that we are in a full control of what events we want to store. Additionally events are easier to query than versions - I believe the only reason we decided to create new partnership instead of updating the existing one in case of reporting previously challenged partnership is for us to have a visibility of how often those challenges happen. While it would still be possible to get this information with papertrail, it would not be as simple as `EventLog.where(owner_type: "Partnership", event: :challenged)`.

Additional benefit of event log is that event log might contain arbitrary information - we can add additional context of the given event - for reporting we could for example include whether it was added via csv upload or school search. Papertrail is concerned with a database record rather than domain object.

In short - papertrail is great when we want to see the state given object was at given time. Event log is a win when we want to query the domain model history.

It is however a good question whether we care enough about this. I think the answer is "not sure", in which case I think the cheap solution like this even log is quite well suited, as we can either expand it scrap it :)

Reply:
I think if the answer is "not sure" then a much cheaper solution is to ask people whose job it is to be sure about (or chase after) those things. I guess Tom might be that person?

Cause it might be that the answer is "we don't care at all from product standpoint, we just want to have an audit trail in case someone tries to cheat". In which case it would be a pity that some minimal work for it `EventLog` has been done, but at least we wouldn't add something we don't need to our codebase. Which truth be told is growing bigger by the day, with a lot of people working on it.

Reply:
Alternatively the answer might be 'yeah, it's great, and we will want a feature like that for admins and analytics people later down the line', and it will stay.

Reply:
@kicferk1 - Very much agree here. Which is basically why I didn't want that audit trail to affect the database design (as we did alter the "simple" structure of unique partnerships to allow auditing). I have found using event logs extremely useful in the past though, especially for support queries when we need to figure out not only what the current state is (or was) but also how and why we got into that state. I think that the transparency given by the log massively outweights the cost of development - it is really just a small addition to this PR, which "meat" really lies in the alteration to the "report_school" form process.

Reply:
The cost of adding an event log in my eyes has little to do with the code change in this PR.

I think the real cost is:
1. We would have two systems that do very similar things (papertrail doesn't deal with 'how' or 'why', sure, but it knows which user got the resource into a particular state, and based on user type we can reverse engineer potential ways into it). 
2. We would need to remember to add relevant event logs everywhere we wanted to track. That adds an extra dimension to code complexity, which needs to be unit tested, which needs to exist in either controllers or services, or forms. One big advantage of papertrail is that it works without the need to add much code, or much tests - but if we want to rely on our EventLog then we probably would need those.
3. Some models will have EventLog, some will have papertrail, some will probably have both, some will have neither. Which imo leads to more developer errors down the line.

Personally, I think these costs outweigh the usefulness, unless we have a product use case for adding them. Bear in mind that when you are the one in control of developing the logging system, you run into a problem of "what was the version of the code when those events got created? And did that version of code log the events I am after?", which is not always easy, and is even harder if we add it to a live system mid-flight as it were. We can maybe do it for the partnerships, but if we try to add them to any other objects, who knows if we can.

Have we ran into a problem (so far) that would be made easy to analyse with EventLogs that was hard to analyse with papertrail? I don't deal with support tickets on R&P so I wouldn't know.

Reply:
> We would have two systems that do very similar things

Actually, the difference between is quite big. Papertrail is a database versioning tool, closely coupled with a database system and as such should be only use for technical purposes. Unfortunately, rails team made at least controversial decision to use ActiveRecord pattern as a domain mode and that decision lead to a number of horrific constructs. While there is nothing wrong with ActiveRecord on its own, using it as a domain object is a commonly analysed anti-pattern (https://softwareengineering.stackexchange.com/questions/119352/does-the-activerecord-pattern-follow-encourage-the-solid-design-principles). Similarly, using papertrail for a domain log can be seen as an anti-pattern. Papertrail is only aware of the state changes, but has no understading of the nature of the change - which is also an artefact of rails resourcefulness. To understand the history of the object, one would need to analyse papertrail of many database tables. On the other hand events are closer to the business logic - we are not updating the record in the database, we are reporting a partnership, which is different to challenging a partnership

> We would need to remember to add relevant event logs everywhere we wanted to track.

This can be seen as a win, as we are in the control of defining the lifecycle of our domain object, decoupled from their database representation.

Few more notes - we do not need a full event log for every object in the system. We are still in the monolithic application and only use the logs to mark important event s in the object lifecycle - shift to a fully blown event driven system would be much more time consuming (however this simple event log would make that shift easier). It is still the database that is the source of truth for a current state. You are correct that we have already lost some of the events and there is no way of retrieving them, which is a shame, but also it means that we should start collecting that data. It is easier to discard data than reconstructing it.

Sorry for being a bit defensive here - there are some very subtle, but important differences here and I guess a lot of personal opinions about the RailsWay of writing the code. Rails has made a number of quite odd design decisions which has triggered an avalanche of problematic patterns and it is quit difficult to get back to "standard way" once you worked outside of those patterns.

I am also not to claim that event log pattern is 100% correct here - the only reason it was introduced was to easily check number of challenges in unit of time as it was explained to me this was the reason for previous, more complex database structure. Introducing event log allows the remaining data structure refactor, without loosing said information. Introducing it here is cheap enough and easy to remove, but I do not want to wait for a business decision before the refactor is done (as it is best when this is merged and deployed before the first challenge occurs).